test/test-rate.js

```diff
@@ -0,0 +1,103 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+const assert = require('assert');
+
+const isClose = require('is-close');
+const rclnodejs = require('../index.js');
+const assertUtils = require('./utils.js');
+const assertThrowsError = assertUtils.assertThrowsError;
+
+const { performance } = require('perf_hooks');
+
+describe('rclnodejs rate test suite', function() {
+  let node;
+  this.timeout(60 * 1000);
+
+  before(function() {
+    return rclnodejs.init();
+  });
+
+  after(function() {
+    rclnodejs.shutdown();
+  });
+
+  beforeEach(function() {
+    node = rclnodejs.createNode('rate_node');
+  });
+
+  afterEach(function() {
+    node.destroy();
+  });
+
+  it('rate constructor tests', function() {
+    const rate1 = node.createRate();
+    assert.equal(rate1.frequency, 1);
+
+    const rate2 = node.createRate(1000);
+    assert.equal(rate2.frequency, 1000);
+
+    const rate3 = node.createRate(0.001);
+    assert.equal(rate3.frequency, 0.001);
+
+    assertThrowsError(() => {
+      node.createRate(1001);
+    }, RangeError);
+
+    assertThrowsError(() => {
+      node.createRate(0);
+    }, RangeError);
+  });
+
+  it('rate api tests', function() {
+    const rate = node.createRate();
+    assert.equal(rate.frequency, 1);
+    assert.equal(rate.isCanceled(), false);
+  });
+
+  it('rate sleep cancel tests', async function() {
+    const rate = node.createRate();
+    assert.equal(rate.isCanceled(), false);
+
+    rate.cancel();
+    assert.ok(rate.isCanceled());
+
+    rate.cancel(); // redundant cancel is a nop
+  });
+
+  it('rate sleep accuracy test, 1000 hz for 3 seconds', async function() {
+    // run 3 * hz,
+    //   collect and average the sleep intervals
+    //   compare average sleep interval with the period of the timer
+    const hz = 1000;
+    const rate = node.createRate(hz);
+
```

Comment:
I am wondering what it's the outcome if I insert:
```js
    setInterval(() => {
      // publish a top here.
    }, 10);
```

Reply:
Here's a simple experiment to see the effect of rate.
Start publishing msgs every 1 sec from commandline using:
```
ros2 topic pub topic 'std_msgs/msg/String' '{data: hello}'
```
Run this example to received msgs every 2 seconds:
```javascript
'use strict';

const rclnodejs = require('./index.js');

async function main() {
  await rclnodejs.init();
  const node = rclnodejs.createNode('test_node');
  const subscriptions = node.createSubscription(
    'std_msgs/msg/String',
    'topic', 
    undefined,
    msg => console.log(`Received(${Date.now()}): ${msg.data}`)
  );
  const rate = node.createRate(0.5);
  let forever = true;
  while (forever) { // should run every 2 secs
    await rate.sleep();
    rclnodejs.spinOnce(node);
  }
}

main();
```

Reply:
Yes, the subscription can work as expected, but publishing a topic can still be called through a callback of the `setInterval()`

Reply:
>but publishing a topic can still be called through a callback of the setInterval()

Correct. Is that not what you were expecting? Are you expecting Rate.sleep() to govern, i.e., block, calls to publish()?

I coded up the example with the setInterval( ()=>publisher.publish('hello'),10). When the interval invokes the callback, then the published msg is available to all subscribers who's node is spinning or spinOnce. In this example  the rate loop is  running at 1/2 hz while a msg is published every 10ms by the setInterval. With that, every 200th msg will be received when spinOnce is called. I believe the 199 msgs received during rate.sleep() are lost.

Reply:
>>but publishing a topic can still be called through a callback of the setInterval()

>Correct. Is that not what you were expecting? Are you expecting Rate.sleep() to govern, i.e., block, calls to publish()?

The behavior (not block the loop) here is reasonable in Node.js, I just want to ensure when developers are calling `sleep()`, they are clear about the expectation (not a real sleep), thanks!